### ISSUE TITLE
Source map and selectors

### ISSUE THREAD

**Issue by unlok** [jakubpawlowicz/clean-css#414]
CSS and source-map from previous step:

### Set 1
css:
```css
a,
a:hover,
a:visited
{
    color: red;
}
```

mappings:
```json
{"version":3,"sources":["index.css"],"names":[],"mappings":"AAAA;;;;IAII,YAAW;EACd"}
```

visual:
```
[a,
a:hover,
a:visited
{
    ][color: red;]
[}]
```

### Set 2

*Like above but selectors written to one line.*

css:
```css
a, a:hover, a:visited
{
    color: red;
}
```

mappings:
```json
{"version":3,"sources":["index.css"],"names":[],"mappings":"AAAA;;IAEI,YAAW;EACd"}
```
visual:
```
[a, a:hover, a:visited
{
    ][color: red;]
[}]
```

I try to load this same sets and optimize it, but correct source-map will generated only for `Set 2`.

For `Set 2` (good mappings).
```js
new CleanCSS({sourceMap: 'form set 2'}).minify('from set 2');
```

mappings:
```json
{"version":3,"sources":["index.css"],"names":[],"mappings":"AAAA,EAAG,QAAS,UAER"}
```

For `Set 1` (bad mappings).
```js
new CleanCSS({sourceMap: 'form set 1'}).minify('from set 1');
```

mappings:
```json
{"version":3,"sources":["index.css","__stdin__.css"],"names":[],"mappings":"AAAA,ECCA,QACA,UDEI"}
```

Why `__stdin__.css` source appeared?

**Comment by unlok** [comment]
@jakubpawlowicz, did you reproduce this problem?

**Comment by jakubpawlowicz** [comment]
@unlok I'll do tomorrow. Sorry, I'm a little behind schedule recently.

**Comment by jakubpawlowicz** [comment]
Ok, I've given it a try and I suspect it may be a bug (?) in postcss. Here's why:

I put the following CSS in source-1.css:
```css
a, a:hover, a:visited
{
    color: red;
}
```

ran it trough the following code

```js
var postcss = require('postcss');
var processor = postcss(function (css, opts) {})
var css = require('fs').readFileSync('source-1.css', 'utf-8');

result = processor.process(css, { map: { inline: false }, from: 'source-1.css', to: 'source-1.processed.css' });
console.log(result.map._mappings._array)
```

The output is as follows:

```js
[ { generatedLine: 1,
    generatedColumn: 0,
    originalLine: 1,
    originalColumn: 0,
    source: 'source-1.css',
    name: null },
  { generatedLine: 3,
    generatedColumn: 4,
    originalLine: 3,
    originalColumn: 4,
    source: 'source-1.css',
    name: null },
  { generatedLine: 3,
    generatedColumn: 16,
    originalLine: 3,
    originalColumn: 15,
    source: 'source-1.css',
    name: null },
  { generatedLine: 4,
    generatedColumn: 2,
    originalLine: 4,
    originalColumn: 1,
    source: 'source-1.css',
    name: null } ]
```

What it's interesting is that `originalLine: 1` is generated only once which means all 3 selectors have same source map position. Clean-css expects every selector to have its own entry in the source map thus the problem with missing mapping.

Now I don't know if that's a problem with postcss or clean-css as I did not find any reference stating either approach is right. Chrome uses more precise position matching (as clean-css) but Firefox uses a less precise one (as postcss).

However I can easily add a fallback so if no correct (?) position is found a closest matching one is used.

Thoughts?

**Comment by unlok** [comment]
Really interesting that the different approaches. Maybe the right way when used minimum mapping if it possible (like **Firefox**)?

But i will be satisfied if you just make fallback for it:)

**Comment by jakubpawlowicz** [comment]
Let's see how the fallback works first!

**Comment by unlok** [comment]
What if add param that **clean-css** will work with source-maps like **post-css**?
```js
sourceMap: {
    inputSource: '...',
    greedy: true
}
```

**Comment by jakubpawlowicz** [comment]
I'd say it should be transparent otherwise people will get confused.

**Comment by jakubpawlowicz** [comment]
@unlok 5dd3e068fc994b12415c7ae9b7d9f1a585310e6b should fix the issue for you - would you be able to give it a try before I push 3.0.5 out?

**Comment by unlok** [comment]
I processed code below through `postcss`
```css
a.button:link,
a.button:visited,
a.button:hover
{
    color: red;
}
```
```json
{
    "version":3,
    "sources":["test.css"],
    "names":[],
    "mappings":"AAAA;;;;IAII,YAAW;EACd"
}
```

Next processed through `cleancss` and again get `__stdin__.css` in mappings.
```css
a.button:hover,a.button:link,a.button:visited{color:red}
```

```json
{
    "version":3,
    "sources":["__stdin__.css","test.css"],
    "names":[],
    "mappings":"AAEA,eCFA,cAAA,iBAII"
}
```

**Comment by jakubpawlowicz** [comment]
I've run it using the latest 3.0 HEAD and got this: https://asciinema.org/a/15493
I presume you used the API but the result should be same as there's even a test with your use case now - https://github.com/jakubpawlowicz/clean-css/commit/5dd3e068fc994b12415c7ae9b7d9f1a585310e6b#diff-9eb831640e84d334de8c721591bfb179R533

Ideas?

**Comment by jakubpawlowicz** [comment]
@unlok I've released it with 3.0.5. Please reopen if needed.

**Comment by unlok** [comment]
@jakubpawlowicz, here's test for my case above - https://github.com/jakubpawlowicz/clean-css/pull/438

**Comment by jakubpawlowicz** [comment]
:+1: I'll pull into a private branch and merge once fixed.

**Comment by jakubpawlowicz** [comment]
@unlok that's fixed in 3.0.6. Hopefully for good this time. Thanks for all the effort testing source maps!

**Comment by unlok** [comment]
@jakubpawlowicz, thanks you too! Good, bug not reproduced. But... :)
If i try to pass source map then result is minified but not merged.

Source:
```css
.test
{
    margin-left: 1px;
    margin-top: 1px;
    margin-bottom: 1px;
    margin-right: 1px;
}
```

```js
new CleanCSS({sourceMap: '...'}).minify(source, function(err, minified) {
    // .test{margin-left:1px;margin-top:1px;margin-bottom:1px;margin-right:1px}
});
```

```js
new CleanCSS().minify(source, function(err, minified) {
    // .test{margin: 1px}
});
```

Or i'm doing something wrong?

**Comment by jakubpawlowicz** [comment]
You do all fine it's just shorthand compacting is disabled when source maps are on - https://github.com/jakubpawlowicz/clean-css/blob/master/lib/clean.js#L43

The only reason behind it is that I haven't figured out what should source map be when we introduce new properties. Any ideas?

**Comment by unlok** [comment]
In this case, may do mapping only for values? And with new properties act as commentary (mapping to nothing).

**Comment by jakubpawlowicz** [comment]
Well if 4 values are turned into one (as in your example) then we should not map it to anything, I guess.

**Comment by unlok** [comment]
Why we can't do mapping for values and their properties separately? Does source-map mapping many-to-one?

![screen shot 2015-01-22 at 15 03 13](https://cloud.githubusercontent.com/assets/3048454/5855305/d1ac3c7a-a247-11e4-9863-2d77fc155dfe.png)

**Comment by unlok** [comment]
Try to load this set on http://evanw.github.io/source-map-visualization/ and see to mapping. It seems normal.

source
```
margin-top: 1px;
margin-right: 1px;
text-align: left;
margin-bottom: 1px;
margin-left: 1px;
```

generated
```
margin: 1px;
```

source-map
```
{"version":3,"sources":["foo.js"],"names":[],"mappings":"AAAA,AACA,AAEA,AACA,MAJU,AACE,AAEC,AACF,EAJC,AACE,AAEC,AACF,IAJG,AACE,AAEC","file":"source-mapped.js"}
```

Idea is that do division of segments by two points instead one.

**Comment by jakubpawlowicz** [comment]
I think it's a good idea indeed. I'll open a new issue just for it.

**Comment by jakubpawlowicz** [comment]
#399 is the one.
